packages/headless/src/controllers/commerce/facets/core/headless-core-commerce-facet.ts

```diff
@@ -0,0 +1,285 @@
+};
+
+export type CoreCommerceFacetBuilder = typeof buildCoreCommerceFacet;
+
+export function buildCoreCommerceFacet<
+  ValueRequest extends AnyFacetValueRequest,
+  ValueResponse extends AnyFacetValueResponse,
+>(
+  engine: CommerceEngine,
+  props: CoreCommerceFacetProps
+): CoreCommerceFacet<ValueRequest, ValueResponse> {
+  if (!loadCommerceFacetReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+
+  const facetId = props.options.facetId;
+
+  const getRequest = () => engine.state.commerceFacetSet[facetId].request;
+  const getResponse = () =>
+    commerceFacetResponseSelector(engine.state, facetId)!;
+  const getIsLoading = () =>
+    isCommerceFacetLoadingResponseSelector(engine.state);
+
+  const getNumberOfActiveValues = () => {
+    return getRequest().values.filter((v) => v.state !== 'idle').length;
+  };
+
+  const computeCanShowLessValues = () => {
+    const request = getRequest();
+    const initialNumberOfValues = request.initialNumberOfValues;
+    const hasIdleValues = !!request.values.find((v) => v.state === 'idle');
+
+    return initialNumberOfValues < request.numberOfValues && hasIdleValues;
+  };
+
+  return {
+    ...controller,
+
+    toggleSelect: (selection: ValueRequest) => {
+      dispatch(props.options.toggleSelectActionCreator({selection, facetId}));
+      dispatch(props.options.fetchResultsActionCreator());
+      // TODO: analytics
+    },
+
+    toggleExclude: (selection: ValueRequest) => {
+      dispatch(props.options.toggleExcludeActionCreator({selection, facetId}));
+      dispatch(props.options.fetchResultsActionCreator());
+      // TODO: analytics
+    },
+
+    // Must use a function here to properly support inheritance with `this`.
+    toggleSingleSelect: function (selection: ValueRequest) {
+      if (selection.state === 'idle') {
+        dispatch(deselectAllFacetValues(facetId));
+      }
+
+      this.toggleSelect(selection);
+    },
+
+    // Must use a function here to properly support inheritance with `this`.
+    toggleSingleExclude: function (selection: ValueRequest) {
+      if (selection.state === 'idle') {
+        dispatch(deselectAllFacetValues(facetId));
+      }
+
+      this.toggleExclude(selection);
+    },
+
+    isValueSelected: (value: ValueResponse) => {
+      return value.state === 'selected';
+    },
+
+    isValueExcluded: (value: ValueResponse) => {
+      return value.state === 'excluded';
+    },
+
+    deselectAll() {
+      dispatch(deselectAllFacetValues(facetId));
+      dispatch(props.options.fetchResultsActionCreator());
+    },
+
+    showMoreValues() {
+      const numberInState = getRequest().numberOfValues;
+      const initialNumberOfValues = getRequest().initialNumberOfValues;
+      const numberToNextMultipleOfConfigured =
+        initialNumberOfValues - (numberInState % initialNumberOfValues);
+      const numberOfValues = numberInState + numberToNextMultipleOfConfigured;
```

Comment:
seems complicated

Reply:
Indeed, this comes from the original core facet so I don't have full context.

I believe the reason why we're not simply doing `numberOfValues = numberInState + initialNumberOfValues` is that the number of values could go down in the response if a value is selected and some values that were previously being displayed no longer have matching results. The % allows us to "reset" the slate to request a nice multiple of the initial number of values.